### PR TITLE
Resolves #1151 - send errors, error-logs and panics to sentry

### LINKED ISSUE #1151: Use structured logging and send it to Sentry
Sentry has been setup already following the instructions in https://forge.rust-lang.org/infra/docs/sentry.html. We could add sentry integration to docs.rs, see https://github.com/rust-lang/crates.io/pull/2961 for an example. There are docs at https://docs.sentry.io/platforms/rust/ and https://docs.rs/sentry/0.20.1/sentry/#quickstart.

Note that sentry has a drop guard for flushing messages, so we should consider implementing #1150 first to avoid losing logs.

### PR DESCRIPTION
**Pull request by syphar**
This PR starts sending errors to sentry. 

What is sent to sentry now (including backtraces): 
- `log::error!` 
- `anyhow::Error`, when ending up in the `main` handler 
- `panic!` everywhere 

I'm not 100% sure if there need to be more places changed? 

### open topics / question 
- should all the `error!` calls in the codebase stay errors being sent to sentry? I could imagine some of them should be changed to warnings? Or should we exclude error-logs from sentry? 
- ~~I would love to switch the handling code in the `ctry!` to use `capture_anyhow` when it gets a fitting error type, but I wasn't sure about the best approach.~~ (done, thanks to @jyn514)

### CODE REVIEW

File: src/bin/cratesfyi.rs
Comment (jyn514) [inline src/bin/cratesfyi.rs:30]: I was reading through the docs and found `environment`: https://docs.sentry.io/platforms/rust/configuration/options/#environment
> Sets the environment. This string is freeform and not set by default. A release can be associated with  more than one environment to separate them in the UI (think staging vs prod or similar).
> By default the SDK will try to read this value from the SENTRY_ENVIRONMENT environment variable (except for the browser SDK where this is not applicable).

Can you set `SENTRY_ENVIRONMENT=dev` in `.env.sample`?

File: src/bin/cratesfyi.rs
Comment (jyn514) [inline src/bin/cratesfyi.rs:53]: This will print the message twice, right? Why did you add it?
```suggestion
```

File: src/build_queue.rs
Comment (jyn514) [inline src/build_queue.rs:124]: Why did you change this? I think sentry will only show the backtrace of the `error!` call, not the backtrace of where the error originally came from.

File: src/bin/cratesfyi.rs
Comment (jyn514) [inline src/bin/cratesfyi.rs:29]: Hmm, why set this? The docs say:
> When enabled, stack traces are automatically attached to all messages logged. Stack traces are always attached to exceptions; however, when this option is set, stack traces are also sent with messages. This option, for instance, means that stack traces appear next to all log messages.

and I don't think it's super useful to show a backtrace for the top-level error! macro, since it will always be the same.

File: src/web/mod.rs
Comment (jyn514) [inline src/web/mod.rs:22]: Why remove this?

File: src/web/mod.rs
Comment (jyn514) [inline src/web/mod.rs:53]: Why remove this?

**Review by jyn514 — COMMENTED** [review]
This is really cool :) do you know how I can test it?

File: src/bin/cratesfyi.rs
Comment (jyn514) [inline src/bin/cratesfyi.rs:51]: Hmm, now that I think about it more I think @Nemo157 had a good point in https://github.com/rust-lang/docs.rs/pull/1152/files#r515400948 - any top-level errors will show up immediately after you run the command, so they'll be something we (me and pietro) can interact with in real time; there's no need to log them. The errors that are important are the ones from `ctry!` and `cexpect!`, which come when the daemon is running unattended. So I don't think we should send an event here.
```suggestion
```

File: src/web/mod.rs
Comment (jyn514) [inline src/web/mod.rs:18]: Let me think about how we could do this, I'll play around with it :)

**Comment by syphar** [comment]
> This is really cool :) do you know how I can test it?

@jyn514 short answer to this question:  
I used my personal sentry account to create a `SENTRY_DSN`, then manually when through the errors to see what ends up in sentry.

File: src/bin/cratesfyi.rs
Comment (syphar) [reply src/bin/cratesfyi.rs:29]: I wanted to have automatic stack-traces for `log::error!` calls. 

I might be wrong, but in testing the stacktrace looked correct, from the place where the `log::error` call was made: 

grafik

( I had an additional `log::error` call in `main`)

Comment (jyn514) [reply src/bin/cratesfyi.rs:28]: > I might be wrong, but in testing the stacktrace looked correct, from the place where the log::error call was made:

Right - but my point is that usually we want the stacktrace of the error itself, not the error! call.

I guess that's an argument for getting rid of error! altogether and changing all those to `capture_anyhow`; I would be ok leaving this in until we do that.

File: src/web/mod.rs
Comment (jyn514) [reply src/web/mod.rs:18]: https://github.com/rust-lang/docs.rs/commit/d6ea893a7829bfb75a1c90e91633641101ee1e65 should make it work once #1489 is merged :)

Comment (jyn514) [reply src/web/mod.rs:18]: And it would be really cool if you could change all the current `log::error!` calls to use `web::report_error` instead - do you mind doing that? Feel free to move that function around and change the interface, I just want to keep the original backtrace :)

File: src/web/mod.rs
Comment (syphar) [reply src/web/mod.rs:22]: it's the same backtrace as the one attached by sentry itself through `attach_stacktrace`. 

Of course we could decide to keep it only for the logs on the server

File: src/web/mod.rs
Comment (syphar) [reply src/web/mod.rs:53]: it's the same backtrace as the one attached by sentry itself through `attach_stacktrace`.

Of course we could decide to keep it only for the logs on the server

File: src/web/mod.rs
Comment (syphar) [reply src/web/mod.rs:18]: @jyn514 good point here, 

I'll dig into generalizing the reporting into one place that we then use everywhere.

File: src/web/mod.rs
Comment (syphar) [reply src/web/mod.rs:22]: with the new macros we will report the original error / its backtrace

**Comment by syphar** [comment]
@jyn514 IMHO you can have another look :) 

I replaced `log::error!` with `report_error` and sometimes `log::error!` with `log::warn!`. 

Right now I'm not sure 
- if all the replacements are ok to replace 
- if the remaining `error!` calls can stay like this 

Also, from a rust perspective, if there is a better way than the `report_error(&anyhow!(err).context(format!(` construct. 
I only can imagine: 
- a single macro for anyhow with context/format? 
- extending std-err so we can call `.context` on it, returning an anyhow-err? ( failure had this, anyhow doesn't)

**Comment by jyn514** [comment]
> if there are duplications where the same error would lead to multiple logs/sentry errors in some cases. we could also just fix it when this happens.

Fixing them as the occur seems fine - I'd rather have too many alerts than too few.

> if there is a better way than the report_error(&anyhow!(err).context(format!( construct.

Another macros seems fine. I don't know how to do it through traits without messing with it on desktop, and I don't think it needs to block the PR.

File: src/build_queue.rs
Comment (Nemo157) [inline src/build_queue.rs:208]: You can move the context up to where the operation happens to avoid the awkwardness here:

```rust
let res = conn.execute(...).with_context(|| {
    format!(
        "error while setting {}-{} to yanked",
        krate.name, krate.version
    )
});
```

File: src/utils/mod.rs
Comment (Nemo157) [reply src/utils/mod.rs:30]: `{:?}` is generally what you should use to print anyhow errors, that will include the context chain and the backtrace.

File: src/build_queue.rs
Comment (syphar) [reply src/build_queue.rs:208]: I'll give it a try later today, from what I remember in many places the code looked awkward after the change.

Comment (Nemo157) [reply src/build_queue.rs:208]: Another option I've been meaning to try out in an `anyhow`-heavy codebase is to use something like [`fn-error-context`](https://docs.rs/fn-error-context) to allow just annotating functions with their context. That may require creating some extra functions though.

File: src/build_queue.rs
Comment (syphar) [reply src/build_queue.rs:132]: I thought these were normal build failures, since I was wrong I'll change it to `report_error`

File: src/storage/s3.rs
Comment (syphar) [reply src/storage/s3.rs:203]: yep, exactly.

Comment (syphar) [reply src/storage/s3.rs:203]: and we don't need the warnings in sentry (yet), the panic is enough

( there is always room for optimization at some point)

File: src/utils/daemon.rs
Comment (syphar) [reply src/utils/daemon.rs:33]: hm... my initial thought was that it's retried anyways in this loop. 

Thinking about it now I see that we would never see an error this way :) 

I'll change it back :)

Comment (syphar) [reply src/utils/daemon.rs:33]: ( or, correctly: I'm changing it to `report_error`)

File: src/utils/queue_builder.rs
Comment (syphar) [reply src/utils/queue_builder.rs:102]: `e`  is `Box<dyn Any + Send>` because `res` is `std::thread::Result`. 
`error!` would at least end up in sentry. 

I don't fully understand the queue logic here, but does the comment (_trying to handle the error won't help_) mean that we should `panic::resume_unwind(e)` here?

File: src/utils/queue_builder.rs
Comment (syphar) [reply src/utils/queue_builder.rs:92]: from what I see this can happen when building the package gives us an `Err`. 
It seems like a normal build failure would result in just an `Ok(false)` from `build_package`. 

So that probably means we should go back to `error!` (or more exactly, `report_error`) here :) 

> Should we also be incrementing the queue attempts here?

Short answer: I don't know since I don't know this part of the codebase yet :)

File: src/build_queue.rs
Comment (syphar) [reply src/build_queue.rs]: I switched to using `.context` and `.with_context` and it doesn't look as bad as I thought it would. 

Let's see what you think :)

**Comment by syphar** [comment]
@jyn514 this is probably ready for another review :)

File: src/utils/queue_builder.rs
Comment (jyn514) [reply src/utils/queue_builder.rs:96]: Hmm, ok - let's wait until we have sentry set up and we can decide whether to increment the attempts based on the errors we see in practice :)

File: src/utils/queue_builder.rs
Comment (jyn514) [reply src/utils/queue_builder.rs]: That comment is about locking the queue (specifically the `expect()`). It's not related to building the package; and indeed I would prefer *not* to panic so the queue will start building again once we unlock it.

**Review by jyn514 — APPROVED** [review]
LGTM :) @Nemo157 feel free to review too if you like, I won't merge this until the next time I do a deploy (so probably no sooner than Saturday).

**Pull request merged**
